packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -410,106 +570,655 @@ class _CupertinoAlertActionSection extends StatelessWidget {
+      // the overall height.
+      return _computeMaxIntrinsicHeightStacked(width);
+    }
+  }
+
+  // Max height of a stack of buttons is the sum of all button heights + a
+  // divider for each button.
+  double _computeMaxIntrinsicHeightStacked(double width) {
+    assert(childCount >= 2);
+
+    final double allDividersHeight = childCount * dividerThickness;
+    return getChildrenAsList().fold(allDividersHeight, (double heightAccum, RenderBox button) {
+      return heightAccum + button.computeMaxIntrinsicHeight(width);
+    });
+  }
+
+  bool _isSingleButtonRow(double width) {
+    if (childCount == 1) {
+      return true;
+    } else if (childCount == 2) {
+      // There are 2 buttons. If they can fit side-by-side then that's what
+      // we want to do. Otherwise, stack them vertically.
+      final double sideBySideWidth = firstChild.computeMaxIntrinsicWidth(double.infinity)
+          + dividerThickness
+          + lastChild.computeMaxIntrinsicWidth(double.infinity);
+      return sideBySideWidth <= width;
+    } else {
+      return false;
+    }
+  }
+
+  @override
+  void performLayout() {
+    if (_isSingleButtonRow(_kCupertinoDialogWidth)) {
+      if (childCount == 1) {
+        // We have 1 button. Our size is the width of the dialog and the height
+        // of the single button.
+        firstChild.layout(
+          constraints,
+          parentUsesSize: true,
+        );
+
+        size = new Size(_kCupertinoDialogWidth, firstChild.size.height + dividerThickness);
+      } else {
+        // Each button gets half the available width, minus a single divider.
+        final BoxConstraints perButtonConstraints = constraints.copyWith(
+          minWidth: (constraints.minWidth - dividerThickness) / 2.0,
+          maxWidth: (constraints.maxWidth - dividerThickness) / 2.0,
+        );
+
+        // Layout the 2 buttons.
+        for(RenderBox button in getChildrenAsList()) {
+          button.layout(
+            perButtonConstraints,
+            parentUsesSize: true,
+          );
+        }
+
+        // The 2nd button needs to be offset to the right.
+        assert(lastChild.parentData is MultiChildLayoutParentData);
+        final MultiChildLayoutParentData secondButtonParentData = lastChild.parentData;
+        secondButtonParentData.offset = new Offset(firstChild.size.width + dividerThickness, 0.0);
+
+        // Calculate our size based on the button sizes.
+        size = new Size(
+          _kCupertinoDialogWidth,
+          math.max(
+            firstChild.size.height,
+            lastChild.size.height,
+          ) + dividerThickness,
+        );
+      }
+    } else {
+      // We need to stack buttons vertically, plus dividers above each button.
+      final BoxConstraints perButtonConstraints = constraints.copyWith(
+        minHeight: 0.0,
+        maxHeight: (constraints.maxHeight - (dividerThickness * childCount)) / childCount,
```

Comment:
i don't understand how this matches the case of >2 buttons, where you show 1.5. If you have 10 buttons and 100 pixels height and each button needs 50 pixels, won't this result in 10 buttons 10 pixels high? (modulo dividerThickness)

Reply:
this needs a test

Reply:
@Hixie I altered the per button constraints to be unbounded for height and then I constrain the overall actions height at the end based on the incoming constraints.

What kind of test were you thinking for this behavior?  There are existing tests for button and content height. There is specifically one titled "Actions section height for 3 buttons is 1.5 buttons tall." That test happens to be incorrectly implemented at the moment, but I'm fixing it.